openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,175 @@
+7. test method names should
+    1. follow the template `should_{expected_behaviour}_when_{state_tested}`
+       e.g. `should_succeed_when_...` & `should_fail_when_...` and use natural language
+    2. not be prefixed by `test_`
+    3. `ValSem` should not be in the test method name but kept in comment
+8. There should not be a `test-utils` feature (best effort). All those helpers should be `#[cfg(test)]` instead and used
+   only in unit tests
+9. Utilities shouldn't unwrap Results to give a chance, depending on the context, to assert an error. Hence, they should
+   all start with `try_` e.g. I might sometimes expect `try_talk_to` to fail even if most of the time I don't.
+10. When it comes to testing MlsGroup, the following utilities would help a lot. More to be added...
+    1. `MlsGroup::try_init(case: &TestCase, client: &str) -> Result(MlsGroup, CredentialBundle)` e.g.
+        ```rust
+        let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+        ```
+    2. `MlsGroup::try_invite<const N: usize>(&mut self, case: &TestCase, others: &mut [&str ; N]) -> Result<[(MlsGroup, CredentialBundle) ; N]>` e.g.
+       ```rust
+       let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+       let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+       ```
+    3. `MlsGroup::try_talk_to<const N: usize>(&mut self, case: &TestCase, others: &mut [MlsGroup ; N])` e.g.
+       ```rust
+       let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+       let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+       assert!(alice_group.try_talk_to(&case, &mut [bob_group, charly_group]).is_ok());
+       ```
+11. Fixtures: in order for being more flexible, they could just require a single `case` variable. It would look like
+    this:
+    ```rust
+    pub struct TestCase<T: OpenMlsCryptoProvider = OpenMlsRustCrypto> {
+        pub ciphersuite: Ciphersuite,
+        pub backend: T,
+        pub cfg: MlsGroupConfig,
+    }
+
+    #[apply(mls_test)]
+    fn should_bla_when_blabla(case: TestCase) {
+        let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+        let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+        assert!(alice_group.try_talk_to(&case, &mut [bob_group, charly_group]).is_ok());
+    }
+    ```
+    1. Does it make sense to keep the backend in fixtures ? I don't see openmls supporting anything else other
+       than `MemoryKeyStore`. Also should we keep evercrypt as an option or discontinue it ?
+    2. `MlsGroupConfig` is introduced to tune the `WireFormatPolicy` to make sure everything also works
+       in `PURE_CIPHERTEXT` mode.
+12. Test file layout
+    * Do not hesitate nesting your tests within mod. They provide better readability. They can be collapsed/expanded.
```

Comment:
Generally there's no need for this. Putting the tests in a module under the file can cover almost all tests for a module and doesn't require collapsing anything.
But tests can of course be added to the file itself if it makes things easier.

Reply:
I am not sure I understand what you mean by that.  
To be more specific, I was thinking about those files with a lot of methods hence a lot of tests e.g. [this one](https://github.com/openmls/openmls/blob/9b1747179a71feaae6cc8ea90be04258cf2aceee/openmls/src/group/mls_group/membership.rs#L15). When writing unit tests for it (so in the same file), I propose the file to look like this:

```rust
impl MlsGroup {
    pub fn add_members();
    pub fn own_leaf();
    pub fn remove_members();
    pub fn propose_add_member();
    pub fn propose_remove_member();
    pub fn leave_group();
    pub fn members();
    pub fn member();
}

#[cfg(test)]
mod tests {
    use super::*;

    mod add_members {
        use super::*;
    }
    mod own_leaf {
        use super::*;
    }
    mod remove_members {
        use super::*;
    }
    mod propose_add_member {
        use super::*;
    }
    mod propose_remove_member {
        use super::*;
    }
    mod leave_group {
        use super::*;
    }
    mod members {
        use super::*;
    }
}
```